### DISCUSSION TITLE
JobStore with external transaction

### DISCUSSION THREAD

**Discussion by justmara** [quartznet/quartznet#1901] [Q&A]
Problem:
I'd like to get a JobStore instance over my already created DB connection - to share the transaction with my business logic. 
Something like:
```
var conn = GetConnection();
var transaction = conn.BeginTransaction();
conn.Execute(..) //some business logic queries
var jobStore = GetQuartzJobStore(conn); // JobStore created over already created connection
jobStore.ScheduleJob(....);
transaction.Comit(); // commits both business code and jobs/triggers creation/modification
```

Currently JobStore is private a property of Scheduler and it is instantiated without our control. Surely one can make IDbProvider that injects some scope-bound DbConnection... But thats still too implicit and fault intolerant.

Variants:
- There is JobStoreCMT that can take usage of TransactionScope, but TransactionScope is an uncontrollable poison itself, so this is not a real deal.
- JobStoreTX uses its own transactions and connections so rollback of business transaction requires manual deletion of jobs/triggers created by quartz.

**Comment by zyu-godaddy** [comment]
Agree. Job status is part of system state, and it should participate in user transactions for overall consistency. We should be able to mark a job as created/completed/failed in user transaction. 

Not sure Quartz is ever going to do that. A workaround is to maintain job status in our own table that's updated transactionally. The job execution method would be something like 

- If the job is not in our status table, it was falsely scheduled (in a transaction that's rolled back). Simply ignore it.
- The job should be 'pending'. Otherwise do not proceed.
- Mark the job as 'in-progress' (atomically with checking 'pending')
- Compute
- Within a transaction, update user data, and mark the job as 'completed'; possibly add more jobs

**Comment by lahma** [comment]
I think it's best to think Quartz as external service from application point of view. One needs to create transactional semantics that don't incur Quartz's job store implementation. 

Might be better to think again about the use case. If it's about preparing data and only if that succeeds should run a job.  If preparation failed - throw an exception and don't schedule job? If data is in the database and job has been scheduled, but something went wrong later - make the job account for invalid starting status - have a grace period before starting the job.

**Reply by justmara** [reply]
But why? Technically all the information JobStore needs - SchedulerId and DbType (to choose the right delegate).
All the triggers/jobs info can be read from db and stored there to be then catched by scheduler's JobStore on db scan.
So a method to create JobStoreCMT instance over provided connection would be enough.

Thinking about quartz as an external resource (while still running over same db) involves pretty complicated and counterintuitive patterns like '_design your jobs like if they always can miss the real data because of reverted business transaction_'. So you could never be sure if it is really reverted transaction or you just messed up JobData on job enqueue and provided wrong id (for example).
